src/main/java/eu/arrowhead/deviceqosevaluator/engine/StatisticsEngine.java

```
@@ -0,0 +1,227 @@
+			final List<StatEntity> data = statDbService.getByDeviceIdAfterTimestamp(metricModel.getGroup(), deviceId, afterTimestamp);
+			if (!hasData(data)) {
+				noStat.add(metricModel.getGroup());
+			}
+
+			for (final Entry<OidMetric, Double> metricEntry : metricModel.getMetricWeight().entrySet()) {
+				score = score + calculateMetricScore(metricEntry.getKey(), metricModel.getScaleTo(), metricEntry.getValue(), data, metricModel.getGroup().getWorstStat());
+			}
+
+		}
+
+		return score;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private boolean hasData(final List<StatEntity> data) {
+		logger.debug("hasData started");
+
+		if (Utilities.isEmpty(data)) {
+			return false;
+		}
+
+		for (final StatEntity stat : data) {
+			if (stat.getMean() != -1) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private double calculateMetricScore(final OidMetric metric, final Double scaleTo, final Double weight, final List<StatEntity> data, final double worstStat) {
+		logger.debug("calculateMetricScore started");
+
+		if (Utilities.isEmpty(data)) {
+			return worstStat * weight;
+		}
+
+		if (metric == OidMetric.CURRENT) {
+			data.sort(Comparator.comparing(StatEntity::getTimestamp)); // ascending
+			return getValue(metric, data.getLast(), scaleTo, worstStat) * weight;
+		}
+
+		final double[] values = new double[data.size()];
+		for (int i = 0; i < values.length; ++i) {
+			values[i] = getValue(metric, data.get(i), scaleTo, worstStat);
+		}
+
+		switch (metric) {
+		case MINIMUM:
+			return Stat.min(values) * weight;
+
+		case MAXIMUM:
+			return Stat.max(values) * weight;
+
+		case MEAN:
+			return Stat.mean(values) * weight;
+
+		case MEDIAN:
+			return Stat.median(values) * weight;
+
+		default:
+			logger.error("Unhandled OID metric: " + metric);
+			return 0;
```

I don't think you should use 0 as an extreme value, because it can be a valid result (for example in case of Minimum)

---

should throw an exception